From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ibramsou <kasobanboui@gmail.com>
Date: Mon, 1 Jul 2024 14:53:12 +0200
Subject: [PATCH] Implement Biome Cache


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 69c12d9049af908380c48c7f13d3d5c7220f8e39..377112cba898bebdd6c19b907e4643681abe37df 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -5,7 +5,6 @@ import com.google.common.base.Predicates;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.mojang.datafixers.util.Pair;
-import it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import java.io.File;
@@ -20,7 +19,6 @@ import java.util.Objects;
 import java.util.Random;
 import java.util.Set;
 import java.util.UUID;
-import java.util.concurrent.ExecutionException;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
@@ -124,7 +122,6 @@ import org.bukkit.entity.TippedArrow;
 import org.bukkit.entity.Trident;
 import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 import org.bukkit.event.weather.LightningStrikeEvent;
-import org.bukkit.event.world.SpawnChangeEvent;
 import org.bukkit.event.world.TimeSkipEvent;
 import org.bukkit.generator.BiomeProvider;
 import org.bukkit.generator.BlockPopulator;
@@ -147,6 +144,7 @@ import org.bukkit.util.StructureSearchResult;
 import org.bukkit.util.Vector;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import javax.annotation.Nonnull;
 
 public class CraftWorld extends CraftRegionAccessor implements World {
     public static final int CUSTOM_DIMENSION_OFFSET = 10;
@@ -2458,6 +2456,25 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     }
     // Purpur end
 
+    @Override
+    public @Nullable Biome getCachedBiome(Chunk chunk) {
+        Holder<net.minecraft.world.level.biome.Biome> holder = this.getHandle().getBiomeManager().getCachedBiome(chunk.getChunkKey());
+        if (holder == null) return null;
+        return CraftBiome.minecraftHolderToBukkit(holder);
+    }
+
+    @Override
+    public @Nullable Biome removeCachedBiome(Chunk chunk) {
+        Holder<net.minecraft.world.level.biome.Biome> holder = this.getHandle().getBiomeManager().removeCachedBiome(chunk.getChunkKey());
+        if (holder == null) return null;
+        return CraftBiome.minecraftHolderToBukkit(holder);
+    }
+
+    @Override
+    public void setCachedBiome(@Nonnull Chunk chunk, @Nonnull Biome biome) {
+        this.getHandle().getBiomeManager().setCachedBiome(chunk.getChunkKey(), CraftBiome.bukkitToMinecraftHolder(biome));
+    }
+
     @Override
     public Collection<GeneratedStructure> getStructures(int x, int z) {
         return this.getStructures(x, z, struct -> true);
